README.md
# HalfFlake

[![Latest Stable Version](https://img.shields.io/packagist/v/real-digital/half-flake.svg)](https://packagist.org/packages/real-digital/half-flake)
[![Build Status](https://travis-ci.org/real-digital/half-flake.svg?branch=master)](https://travis-ci.org/real-digital/half-flake)
[![Coverage Status](https://coveralls.io/repos/github/real-digital/half-flake/badge.svg?branch=master)](https://coveralls.io/github/real-digital/half-flake?branch=master)

HalfFlake is a PHP library for distributive generating unique ID numbers using
Twitter's [Snowflake](https://github.com/twitter-archive/snowflake/blob/snowflake-2010/README.mkd) Algorithm.

## Installation

via Composer

``` bash
$ composer require real-digital/half-flake
```

via GitHub

``` bash
$ git clone https://github.com/real-digital/half-flake.git
```

### Usage

```php
<?php

use Real\HalfFlake;

// an unique pair of constants defined per node
const ID_DATACENTER = 1;
const ID_NODE = 1;

$shard = new HalfFlake\Seed(ID_DATACENTER, ID_NODE);
$clock = new HalfFlake\Time();

try {
    $generator = new HalfFlake\Generator($shard, $clock);
    $id = $generator->nextId();
} catch (HalfFlake\RuntimeException $e) {
    // handle errors
}
```

## Testing

```bash
composer test
```

## License

HalfFlake is licensed under the MIT License. Please see [LICENSE](LICENSE) for details.


## Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information.


CHANGELOG.md
[1.0.0]: https://github.com/real-digital/half-flake/commits/1.0.0

# Changelog

All notable changes to this project will be listed in this file.

## [1.0.0] - 2019-06-27

The first release as a separated library


tests/SeedTest.php
<?php
declare(strict_types=1);

namespace Real\HalfFlake\Tests;

use PHPUnit\Framework\TestCase;
use Real\HalfFlake;

class SeedTest extends TestCase
{
    /** @var HalfFlake\Seed */
    private $seed;

    protected function setUp(): void
    {
        parent::setUp();

        $this->seed = new HalfFlake\Seed(2, 3);
    }

    public function testDatacenter(): void
    {
        self::assertSame(2, $this->seed->datacenter());
    }

    public function testNode(): void
    {
        self::assertSame(3, $this->seed->node());
    }
}


tests/TimeTest.php
<?php
declare(strict_types=1);

namespace Real\HalfFlake\Tests;

use PHPUnit\Framework\TestCase;
use Real\HalfFlake;

/**
 * @runTestsInSeparateProcesses
 * @preserveGlobalState disabled
 */
class TimeTest extends TestCase
{
    private const NOW = 330;

    /**
     * We just overloading built-in function in a origin namespace to mock the clock
     *
     * @see setUP()
     * @link http://php.net/manual/en/language.namespaces.fallback.php
     * @link https://github.com/symfony/phpunit-bridge/blob/master/ClockMock.php
     */
    private const CLOCK_MOCK = <<< EO_CLOCK_MOCK
        namespace Real\HalfFlake;
        
        function microtime(\$asFloat = false)
        {
            return %d;
        }
EO_CLOCK_MOCK;

    public function setUp(): void
    {
        parent::setUp();

        eval(sprintf(self::CLOCK_MOCK, self::NOW));
    }

    public function testNow(): void
    {
        $time = new HalfFlake\Time();

        self::assertSame(self::NOW, $time->now()->getTimestamp());
    }

    public function testEpoch(): void
    {
        $epochOffset = strtotime(HalfFlake\Time::EPOCH_BEGINNING) * 1000;

        $time = new HalfFlake\Time();

        self::assertSame(self::NOW * 1000, $epochOffset + $time->epoch());
    }
}


tests/GeneratorTest.php
<?php
declare(strict_types=1);

namespace Real\HalfFlake\Tests;

use PHPUnit\Framework\TestCase;
use Real\HalfFlake;

class GeneratorTest extends TestCase
{
    public function testGeneratorInterfaceIsInherited(): void
    {
        $shard = $this->createConfiguredMock(HalfFlake\Shard::class, [
            'datacenter' => 1,
            'node' => 1,
        ]);

        $clock = $this->createMock(HalfFlake\Clock::class);
        $generator = new HalfFlake\Generator($shard, $clock);

        self::assertInstanceOf(HalfFlake\Algorithm::class, $generator);
        self::assertInstanceOf(HalfFlake\Algorithm::class, $generator);
    }

    public function invalidSeedProvider(): array
    {
        return [
            [0, 0, 1000],
            [0, 1, 1000],
            [33, 1, 1000],
            [1, 0, 1001],
            [1, 65, 1001],
        ];
    }

    /**
     * @dataProvider invalidSeedProvider
     */
    public function testInvalidShardThrowsException(int $datacenter, int $node, int $exceptionCode): void
    {
        $clock = $this->createMock(HalfFlake\Clock::class);

        $shard = $this->createConfiguredMock(HalfFlake\Shard::class, [
            'datacenter' => $datacenter,
            'node' => $node,
        ]);

        $this->expectException(HalfFlake\RuntimeException::class);
        $this->expectExceptionCode($exceptionCode);

        new HalfFlake\Generator($shard, $clock);
    }

    public function testSequenceIsIncrementing(): void
    {
        $shard = $this->createConfiguredMock(HalfFlake\Shard::class, [
            'datacenter' => 1,
            'node' => 1,
        ]);

        $clock = $this->createMock(HalfFlake\Clock::class);
        $generator = new HalfFlake\Generator($shard, $clock);

        // half of the range because we call nextId twice
        $iteration = (1 << HalfFlake\Algorithm::BITS_SEQUENCE) / 2;

        while ($iteration > 0) {
            --$iteration;
            $clock->method('epoch')->will($this->returnValue($iteration));

            self::assertNotSame(
                $generator->nextId(),
                $generator->nextId(),
                sprintf('Error on iteration #%d', $iteration)
            );
        }

        // next id must trigger sequence overflow
        $this->expectException(HalfFlake\RuntimeException::class);
        $this->expectExceptionCode(1002);

        $generator->nextId();
    }

    public function equalShardIdentifierProvider(): array
    {
        return [
            [1, 2],
            [2, 3],
            [3, 4],
            [1, 5],
            [3, 6],
        ];
    }

    /**
     * @dataProvider equalShardIdentifierProvider
     */
    public function testDifferentIdDatacenterWithSameIdNodeGeneratesDifferentIds(
        int $idDatacenter1,
        int $idDatacenter2
    ): void {
        $shard1 = $this->createConfiguredMock(HalfFlake\Shard::class, [
            'datacenter' => $idDatacenter1,
            'node' => 2,
        ]);

        $shard2 = $this->createConfiguredMock(HalfFlake\Shard::class, [
            'datacenter' => $idDatacenter2,
            'node' => 2,
        ]);

        $clock = $this->createConfiguredMock(HalfFlake\Clock::class, ['epoch' => 1]);

        $generator1 = new HalfFlake\Generator($shard1, $clock);
        $generator2 = new HalfFlake\Generator($shard2, $clock);

        self::assertNotSame($generator1->nextId(), $generator2->nextId());
    }

    /**
     * @dataProvider equalShardIdentifierProvider
     */
    public function testDifferentIdNodeWithinOneDatacenterGeneratesDifferentIds(int $idNode1, int $idNode2): void
    {
        $shard1 = $this->createConfiguredMock(HalfFlake\Shard::class, [
            'datacenter' => 2,
            'node' => $idNode1,
        ]);

        $shard2 = $this->createConfiguredMock(HalfFlake\Shard::class, [
            'datacenter' => 2,
            'node' => $idNode2,
        ]);

        $clock = $this->createConfiguredMock(HalfFlake\Clock::class, ['epoch' => 1]);

        $generator1 = new HalfFlake\Generator($shard1, $clock);
        $generator2 = new HalfFlake\Generator($shard2, $clock);

        self::assertNotSame($generator1->nextId(), $generator2->nextId());
    }

    public function timestampProvider(): array
    {
        return [
            [-1],
            [4398046511104], //max decimal exactly 42 bits long
        ];
    }

    /**
     * @dataProvider timestampProvider
     */
    public function testTimestampOverflow(int $timestamp): void
    {
        $shard = $this->createConfiguredMock(HalfFlake\Shard::class, [
            'datacenter' => 1,
            'node' => 1,
        ]);

        $clock = $this->createConfiguredMock(HalfFlake\Clock::class, ['epoch' => $timestamp]);

        $generator = new HalfFlake\Generator($shard, $clock);

        $this->expectException(HalfFlake\RuntimeException::class);
        $this->expectExceptionCode(1003);

        $generator->nextId();
    }

    public function testClockMovedBackwards(): void
    {
        $shard = $this->createConfiguredMock(HalfFlake\Shard::class, [
            'datacenter' => 1,
            'node' => 1,
        ]);

        $clock = $this->createMock(HalfFlake\Clock::class);
        $clock->expects($this->any())
            ->method('epoch')
            ->will($this->onConsecutiveCalls(42, 41));

        $generator = new HalfFlake\Generator($shard, $clock);

        $generator->nextId();

        $this->expectException(HalfFlake\RuntimeException::class);
        $this->expectExceptionCode(1004);

        $generator->nextId();
    }
}


src/Generator.php
<?php
declare(strict_types=1);

namespace Real\HalfFlake;

final class Generator implements Algorithm
{
    /**
     * Since PHP does not support unsigned integers and negative IDs are something weird,
     * we are going to limit the IDs capacity within range [1; PHP_INT_MAX]
     *
     * @see PHP_INT_MAX
     * @see Algorithm::BITS_SIGN
     */
    private const SIGN = 0;

    private const SHIFT_SIGN = self::BITS_TOTAL - self::BITS_SIGN;
    private const SHIFT_TIMESTAMP = self::SHIFT_SIGN - self::BITS_TIMESTAMP;
    private const SHIFT_DATACENTER = self::SHIFT_TIMESTAMP - self::BITS_DATACENTER;
    private const SHIFT_NODE = self::SHIFT_DATACENTER - self::BITS_NODE;
    private const SHIFT_SEQUENCE = self::SHIFT_NODE - self::BITS_SEQUENCE;

    /** @var int */
    private $datacenter;

    /** @var int */
    private $node;

    /** @var Clock */
    private $clock;

    /** @var int */
    private $lastTimestamp = 0;

    /** @var int */
    private $sequence = 0;

    public function __construct(Shard $shard, Clock $clock)
    {
        $datacenter = $shard->datacenter();

        if ($datacenter < 1 || $datacenter > (1 << Algorithm::BITS_DATACENTER)) {
            throw new RuntimeException('Invalid datacenter', RuntimeException::CODE_DATACENTER);
        }

        $node = $shard->node();

        if ($node < 1 || $node > (1 << Algorithm::BITS_NODE)) {
            throw new RuntimeException('Invalid node', RuntimeException::CODE_NODE);
        }

        $this->datacenter = $datacenter - 1;
        $this->node = $node - 1;

        $this->clock = $clock;
    }

    /**
     * @return int shifted timestamp with a milliseconds precision
     * @throws RuntimeException
     */
    private function timestamp(): int
    {
        $timestamp = $this->clock->epoch();

        if ($timestamp < 0 || $timestamp >= (1 << self::BITS_TIMESTAMP)) {
            throw new RuntimeException('Timestamp overflow!', RuntimeException::CODE_TIMESTAMP_OVERFLOW);
        }

        if ($timestamp < $this->lastTimestamp) {
            throw new RuntimeException('Clock moved backwards!', RuntimeException::CODE_CLOCK_BACKWARD);
        }

        return $timestamp;
    }

    /**
     * @return int the next sequence number within the same millisecond
     * @throws RuntimeException
     */
    private function sequence(int $timestamp): int
    {
        if ($timestamp === $this->lastTimestamp) { //increment
            ++$this->sequence;

            if ($this->sequence >= (1 << self::BITS_SEQUENCE)) {
                throw new RuntimeException('Generating too fast, slow down!', RuntimeException::CODE_SEQUENCE_OVERFLOW);
            }
        } else { //reset
            $this->sequence = 0;
        }

        $this->lastTimestamp = $timestamp;

        return $this->sequence;
    }

    private function generate(): int
    {
        $timestamp = $this->timestamp();
        $sequence = $this->sequence($timestamp);

        $id = self::SIGN << self::SHIFT_SIGN;
        $id |= $timestamp << self::SHIFT_TIMESTAMP;
        $id |= $this->datacenter << self::SHIFT_DATACENTER;
        $id |= $this->node << self::SHIFT_NODE;
        $id |= $sequence << self::SHIFT_SEQUENCE;

        return $id;
    }

    /**
     * @inheritdoc
     */
    public function nextId(): string
    {
        return (string)$this->generate();
    }
}


src/Algorithm.php
<?php
declare(strict_types=1);

namespace Real\HalfFlake;

/**
 * This Interface defines Twitter's Snowflake algorithm basics terms
 *
 * @link https://github.com/twitter/snowflake/blob/snowflake-2010/README.mkd
 */
interface Algorithm
{
    /** BigInt capacity */
    public const BITS_TOTAL = 64;

    /**
     * Since PHP does not support unsigned integers,
     * take care of implementations written on languages other than PHP having native unsigned integers support
     */
    public const BITS_SIGN = 1;

    /**
     * Timestamp with a milliseconds precision (41 bits) with an additional bit (+1 bit)
     * to postpone the Year 2038 problem @link https://en.wikipedia.org/wiki/Year_2038_problem
     */
    public const BITS_TIMESTAMP = 42;

    /** Decimal max 32 */
    public const BITS_DATACENTER = 5;

    /** Decimal max 64 */
    public const BITS_NODE = 6;

    /** Decimal max 1024 IDs per shard per millisecond */
    public const BITS_SEQUENCE = 10;

    /**
     * Generate next identity
     */
    public function nextId(): string;
}
